protocol/x/clob/keeper/process_operations.go

```diff
@@ -248,6 +214,218 @@ func (k Keeper) PersistMatchToState(
+		// parameter rather than the subaccountOpenOrders map.
+		_, successPerSubaccountUpdate := k.AddOrderToOrderbookCollatCheck(
+			ctx,
+			orderToRemove.GetClobPairId(),
+			subaccountOpenOrders,
+		)
+		if successPerSubaccountUpdate[orderIdToRemove.SubaccountId].IsSuccess() {
+			return errorsmod.Wrapf(
+				types.ErrInvalidOrderRemoval,
+				"Order Removal (%+v) invalid. Order passes collateralization check.",
+				*orderRemoval,
+			)
+		}
+	case types.OrderRemoval_REMOVAL_REASON_INVALID_REDUCE_ONLY:
+		if !orderToRemove.IsReduceOnly() {
+			return errorsmod.Wrapf(
+				types.ErrInvalidOrderRemoval,
+				"Order Removal (%+v) invalid. Order must be reduce only.",
+				*orderRemoval,
+			)
+		}
+
+		// The reduce-only order must increase or change the side of the position to trigger removal.
+		currentPositionSize := k.GetStatePosition(
+			ctx,
+			orderIdToRemove.SubaccountId,
+			orderToRemove.GetClobPairId(),
+		)
+		orderQuantumsToFill := orderToRemove.GetBigQuantums()
+
+		orderFillWouldIncreasePositionSize := orderQuantumsToFill.Sign() == currentPositionSize.Sign()
+
+		newPositionSize := new(big.Int).Add(currentPositionSize, orderQuantumsToFill)
+		orderChangedSide := currentPositionSize.Sign()*newPositionSize.Sign() == -1
+		if !orderFillWouldIncreasePositionSize && !orderChangedSide {
+			return errorsmod.Wrapf(
+				types.ErrInvalidOrderRemoval,
+				"Order Removal (%+v) invalid. Order fill must increase position size or change side.",
+				*orderRemoval,
+			)
+		}
+	case types.OrderRemoval_REMOVAL_REASON_POST_ONLY_WOULD_CROSS_MAKER_ORDER:
+		// TODO
+		//
+		// This removal reason implies the stateful taker order could not be placed on the book because it crossed
+		// some unspecified maker order. We must verify that the stateful taker order should have been placed in this block
+		// and that the maker order it crossed with actually exists on the proposer's orderbook.
+
+		// The order should be post-only
+		if orderToRemove.TimeInForce != types.Order_TIME_IN_FORCE_POST_ONLY {
+			return errorsmod.Wrapf(
+				types.ErrInvalidOrderRemoval,
+				"Order Removal (%+v) invalid. Order is not post-only.",
+				*orderRemoval,
+			)
+		}
+	case types.OrderRemoval_REMOVAL_REASON_INVALID_SELF_TRADE:
+		// TODO
+		//
+		// This removal reason implies the maker order was removed during a self-trade encountered during matching.
+		// We must verify that the stateful order had already been placed, and that the taker order matching which caused
+		// this removal would have matched against this order.
+	case types.OrderRemoval_REMOVAL_REASON_CONDITIONAL_FOK_COULD_NOT_BE_FULLY_FILLED:
+		// TODO: The order must have been triggered in the last block
+
+		// The order should be in triggered state.
+		_, found := k.GetTriggeredConditionalOrderPlacement(ctx, orderIdToRemove)
+		if !found {
+			return errorsmod.Wrapf(
+				types.ErrInvalidOrderRemoval,
+				"Order Removal (%+v) invalid. Order is not in triggered state.",
```

Comment:
nit: instead of calling errorsmod.Wrapf for the errors within this function, simply return the error, e.g. types.ErrOrderNotTriggered, unless you have additional information to apply.

the caller of `PersistOrderRemovalToState` can do the Wrapf and provide additional details (in this case %+v order removal). This is the point of chained errors, so that each layer can provide some additional information